Split RzAnalysis.reg and RzDebug.reg

thestr4ng3r opened the pull request:
# DO NOT SQUASH THIS!!!

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

The core of these changes is removing this single piece of code:
https://github.com/rizinorg/rizin/blob/96a85e573e766e3870a01482f36c92df403cc4cd/librz/debug/dreg.c#L167-L170

This would make `analysis->reg` and `dbg->reg` point to the same `RzReg` which causes multiple issues, for example:
* Even though both debug and analysis plugins supply register profiles, only one could be used at a time. If the debug one was loaded (which it must be as long as debug is active) and at some point e.g. an `analysis.bits` change was triggered, the analysis one was loaded and debug registers were completely messed up because the plugin supplies data only for its own profile.
* Emulation-based analysis can leak into the debuggee. This was partially addressed by #2004 but would still be dangerous.

Unfortunately, pretty much all code was written with this behavior in place. The entire rest of this pr is for fixing these things. I went through every single access of RzDebug.reg and checked it. Hopefully nothing was missed.

**Test plan**

Try out anything that accesses registers in both debug and non-debug/emulation mode. Help on testing here is welcome, especially for debug.

# DO NOT SQUASH THIS!!!

thestr4ng3r committed:
Refactor debug checks and flags from registers

* Added rz_core_is_debug() to check for debug mode instead of hard-coding "cfg.debug"
* rz_core_reg_default() uses the above to get the currently relevant RzReg
* rz_core_debug_regs2flags() -> rz_core_reg_update_flags() uses the above
  to update the flags.

thestr4ng3r committed:
Prepare various RzReg accesses for splitting

Common accesses for both debug and analysis must not use
debug->reg or analysis->reg, but go through rz_core_reg_default().

thestr4ng3r committed:
Split RzDebug.reg and RzAnalysis.reg

The two instances of RzReg were assigned to the same pointer in core.
This had several disadvantages, in particular:
* Both debug and analysis supply reg profiles, which might not be
  compatible. Which one was used in practice depended on which eval
  vars were set in which order.
* Emulation, in particular esil-based analysis, could leak into the
  debuggee.
Now they are fully independent.

thestr4ng3r committed:
Make Core ESIL register accesses point to Analysis

thestr4ng3r committed:
Refactor rz_debug_reg_get()/set()

* Logic for getting a register by role or name, whichever matches, moved
  to RzReg (rz_reg_get_by_role_or_name)
* Removed rz_debug_reg_err() because it's unnecessary
* Added rz_core_reg_getv_by_role_or_name() and
  rz_core_reg_aetv_by_role_or_name() to replace rz_debug_reg_get() for
  for usage whenever it is not debug-specific.

thestr4ng3r committed:
Make as command ready for RzReg splitting

* rz_debug_arg_get/set moved to core as rz_core_arg_get
* Syscall (as) code refactored to always use rz_core_reg_default()
  instead of hardcoded debug

ret2libc commented on librz/core/creg.c:
```diff
@@ -0,0 +1,96 @@
+	return rz_core_is_debug(core) ? core->dbg->reg : core->analysis->reg;
+}
+
+/// rz_reg_getv_by_role_or_name() on rz_core_reg_default()
```
is `///` alright?

shouldn't it be:
```C
/**
  * xxxxxx
  */
```
?

thestr4ng3r replied:
Just re-checked and you are right. Seems like I have been doing this wrong for quite a while...

thestr4ng3r replied:
To be exact: just using `///` works if there are at least two lines. With one it does not. Documented here: https://www.doxygen.nl/manual/docblocks.html#cppblock

ret2libc commented on librz/core/creg.c:
```diff
@@ -0,0 +1,96 @@
+
+/// set on rz_core_reg_default()
+RZ_API bool rz_core_reg_set_by_role_or_name(RzCore *core, const char *name, ut64 num) {
+	// this logic has to be in sync with rz_core_reg_default().
+	if (rz_core_is_debug(core)) {
+		// call this instead of rz_reg_getv_... directly because it also syncs
+		return rz_debug_reg_set(core->dbg, name, num);
+	}
+	RzRegItem *ri = rz_reg_get_by_role_or_name(core->analysis->reg, name);
+	if (!ri) {
+		return false;
+	}
+	return rz_reg_set_value(core->analysis->reg, ri, num);
```
shouldn't this become:
```C
reg = rz_core_reg_default(core);
RzRegItem *ri = rz_reg_get_by_role_or_name(reg, name);
return rz_reg_set_value(reg, ri, num);
```
?

thestr4ng3r replied:
It is ok like this because `rz_core_is_debug(core)` has already been checked. This is what the comment `this logic has to be in sync with rz_core_reg_default()` means.

ret2libc replied:
Yes but why do you want to sync regs in one case and not in the other?

thestr4ng3r replied:
It's only for syncing with the debuggee. In the case of analysis->regs, there is no such thing.

ret2libc replied:
Don't you need to sync the registers with the flags anyway?

thestr4ng3r replied:
Theoretically yes, but flags are completely separate, it's done by the caller and not here. In general I am not a fan at all of all the sync calls for both debug and flags. At least flags could be done much better with events, but that is another issue.

ret2libc replied:
I'm just not sure about this because callers of this function would have different visible results based on debug vs non-debug mode. In debug mode they can expect flags to be synced, in non-debug mode they can't. That means that the caller needs to be aware of what this function does, which counters a bit the purpose of having a unified function IMO.

thestr4ng3r replied:
No, flags are not synced at all in this function, regardless of the mode.

ret2libc replied:
Oh, I thought `because it also syncs` was referring to the flags, sorry

thestr4ng3r replied:
After looking at more code that will use this, I have changed my mind about syncing flags. It now also syncs the flags and that happens in both debug and emulation mode.

ret2libc reviewed (commented):
remember to add doxygen doc to new RZ_API functions before marking this PR ready.

thestr4ng3r committed:
Make fs+ trim whitespace of the arg

Previously, `fs+ myname` would use the flagspace " myname" with a
leading space. This made .arf inconsistent with setting regs-flags from
C and the syntax `fs+myname` would not be allowed in the future with
newshell anyway.

thestr4ng3r committed:
Remove RzDebug.creg and clean its usages

This was holding the name of a "current" register that was awkwardly set
during register printing, to then be used for setting in visual mode.
The features make sense, but it's too confusingly implemented to be
useful or stable.

thestr4ng3r committed:
Add basic tests for de (ESIL watchpoints)

wargio commented on librz/core/carg.c:
```diff
@@ -5,6 +5,70 @@
+		}
+	}
+	char reg[32];
+	snprintf(reg, sizeof(reg) - 1, "A%d", num);
```
rz_strf

wargio commented on librz/core/carg.c:
```diff
@@ -5,6 +5,70 @@
+		return true;
+	}
+	char reg[32];
+	snprintf(reg, sizeof(reg) - 1, "A%d", num);
```
same here.
